server/servers/api/src/main/scala/com/prisma/api/schema/Optimizations.scala

```diff
@@ -0,0 +1,126 @@
+package com.prisma.api.schema
+
+import com.prisma.api.connector._
+
+object Optimizations {
+
+  trait Optimizer {
+    def optimize(filter: Filter): Filter
+
+  }
+
+  object FilterOptimizer extends Optimizer {
+
+    override def optimize(filter: Filter): Filter = {
+      val one = LogicalOpt.transform(filter)
+//      val two   = InlineOpt.transform(one)
+      val three = SameRelationFilterOpt.transform(one)
```

Comment:
To avoid naming all those intermediate variables you could put all the `.transform` functions into an an array and use a `foldLeft` to apply all of them.

Reply:
done